### PR TITLE
Set Support : add method

### LINKED ISSUE #1738: add : Adds an element to the set
### Feature Description

**Add support for add() method for sets in the Python translation.**
`add(element)` : Adds an element to the set if it's not already present. 
If the element is already in the set, the set remains unchanged.

### Test Code

This Code Should Work 

```python
my_set = {1, 2, 3}

my_set.add(4)

print(my_set) 
>> {1, 2, 3, 4}
```

### Proposed Solution
See description in #1531

### PR DESCRIPTION
**Pull request by mustapha-belbiad**
Add support for the add() method of Python sets to the semantic stage. 
Add handling of that method to the Python printer. This fixes  #1738 

**Commit Summary**

- Add a class named `NativeHomogeneousSet` to represent the homogenous set type in `pyccel.ast.datatypes`.
- Add a `PythonSet` class to represent a call to Python's {,} in `pyccel.ast.builtins`.
- In `pyccel.parser.syntactic`, add _visit_Set that calls the `pythonSet` class.
- Add `_visit_PythonSet` in `pyccel.parser.semantic` and `_print_PythonSet` to the Python printer for generating appropriate Python code.
- Create a file named set_methods in `pyccel.ast.builtin_methods`.
- Add a class representing the `add()` method, inheriting from `PyccelInternalFunction`.
- Create a ClassDef class representing a `NativeHomogeneousSet`. This class will include an `add()` method implemented by the `SetAdd` class.
- Update the condition in the `get_cls_base()` function to properly check the class type of the containers.
- Add the `_print_Setadd()` method to the Python printer for generating the appropriate Python code for the `add()` method.
- Add tests for the `add()` method involving various scenarios where the function could be used.

### CODE REVIEW

File: pyccel/ast/builtins.py [hunk 1]
```diff
@@ -770,6 +771,72 @@ def is_homogeneous(self):
...
+    @property
+    def is_homogeneous(self):
+        """
+        Indicates whether the set is homogeneous or inhomogeneous.
+
+        Indicates whether all elements of the Set have the same dtype, precision,
+        rank, etc (homogenous) or if these values can vary (inhomogeneous). sets
+        are always homogeneous.
+        """
+        return True
```
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/ast/builtins.py:838]: /bot accept 
The sets will always be homogeneous, so it's just return True.

File: pyccel/ast/datatypes.py [hunk 2]
```diff
@@ -331,6 +331,23 @@
...
+
+    Class representing the type of a homogeneous Set. This
+    is a container type and should be used as the class_type.
+    """
+    __slots__ = ()
+    _name = 'Set'
+
+    @lru_cache
+    def __add__(self, other):
+        if isinstance(other, NativeHomogeneousSet):
```
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/ast/datatypes.py:347]: /bot accept
This method defines behavior for the addition operator `+`.

File: pyccel/parser/semantic.py [hunk 3]
```diff
@@ -2198,6 +2198,18 @@ def _visit_PythonList(self, expr):
                 severity='fatal')
         return expr
 
+    def _visit_PythonSet(self, expr):
+        ls = [self._visit(i) for i in expr]
+        try:
+            expr = PythonSet(*ls)
+        except TypeError as e:
+            message = str(e)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:2206]: I agree :+1:

File: pyccel/ast/datatypes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/datatypes.py:347]: You are correct that this method defines behaviour for the addition operator `+`. But I don't think that that is a reason to add code here without a test. You should beware, this method does **not** define the behaviour when adding two sets it refers only to the datatypes. The addition of 2 datatypes does not have a logical meaning but is simply a method that we use to quickly deduce types inside operators.

I am in the process of rewriting this file. The new docstring which describes this method reads:
> Types contain an addition operator. The operator indicates the type that
> is expected when calling an arithmetic operator on objects of these types.

So this should describe what happens if I do `a+b` or `a-b` or `a*b`.

What code where you hoping to describe with this operator?

File: pyccel/ast/builtin_methods/set_methods.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+__all__ = ('SetAdd',)
+
+class SetAdd(PyccelInternalFunction) :
+    """
+    Represents a call to the .add() method.
+    
+    Represents a call to the .add() method which add an element
+    to the set if it's not already present.
+    If the element is already in the set, the set remains unchanged.
+    Parameters
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_methods/set_methods.py:25]:
```suggestion
    If the element is already in the set, the set remains unchanged.

    Parameters
```

Comment (EmilyBourne) [reply pyccel/ast/builtin_methods/set_methods.py:25]: Numpydoc cannot detect the parameters section without this space

Comment (mustapha-belbiad) [reply pyccel/ast/builtin_methods/set_methods.py]: fixed : d7839d8

File: pyccel/ast/datatypes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/ast/datatypes.py:347]: i think this code will do : 
```Python
se1 = {(2, 4, 5)}
se2 = {(7, 8, 1)}
result = se1 + se2
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/datatypes.py:347]: Almost. This code is indeed designed to find the type of `result` but:
image

[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/ast/datatypes.py:347]: what about : 
```Python
se1 = {(2, 4, 5)}
se2 = {(7, 8, 1)}
result = se1 | se2
```
It may not be  `+`, but it still functions similarly and serves as a shortcut for the ` union()` method.

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/datatypes.py:347]: `|` is a bitwise operator so it doesn't use the `__add__` method which is designed for arithmetic operators

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/datatypes.py:347]: This is similar to `+` for lists which is handled here:
https://github.com/pyccel/pyccel/blob/f42808b62d57a3f53787491cd0a22d62b9928031/pyccel/parser/semantic.py#L2679-L2704

File: pyccel/codegen/printing/pycode.py [hunk 5]
```diff
@@ -442,6 +442,10 @@ def _print_PythonList(self, expr):
         args = ', '.join(self._print(i) for i in expr.args)
         return '['+args+']'
 
+    def _print_PythonSet(self, expr):
+        args = ', '.join(self._print(i) for i in expr.args)
+        return '{'+args+'}'
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/pycode.py:447]: This will not work if you want to create an empty set. You will print `{}` but you need to print `set()` or `{,}`

File: pyccel/errors/messages.py [hunk 6]
```diff
@@ -111,6 +111,7 @@
 PYCCEL_RESTRICTION_LIST_COMPREHENSION_SIZE = 'Could not deduce the size of this list comprehension. If you believe this expression is simple then please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem.'
 PYCCEL_RESTRICTION_LIST_COMPREHENSION_LIMITS = 'Pyccel cannot handle this list comprehension. This is because there are occasions where the upper bound is smaller than the lower bound for variable {}'
 PYCCEL_RESTRICTION_INHOMOG_LIST = 'Inhomogeneous lists are not supported by Pyccel. Please use a tuple'
+PYCCEL_RESTRICTION_INHOMOG_SET = 'Inhomogeneous Sets are not supported by Pyccel. Please use a tuple'
```
Comment (EmilyBourne) [inline pyccel/errors/messages.py:114]: How can you use a tuple in place of a set?

File: tests/epyccel/test_epyccel_sets.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# pylint: disable=missing-function-docstring, missing-module-docstring
+import pytest
+from  pyccel.epyccel import epyccel
+
+@pytest.fixture( params=[
+        pytest.param("fortran", marks = [
+            pytest.mark.skip(reason="list methods not implemented in fortran"),
+            pytest.mark.fortran]),
+        pytest.param("c", marks = [
+            pytest.mark.skip(reason="list methods not implemented in c"),
```
Comment (EmilyBourne) [inline tests/epyccel/test_epyccel_sets.py:10]:
```suggestion
            pytest.mark.skip(reason="set methods not implemented in fortran"),
            pytest.mark.fortran]),
        pytest.param("c", marks = [
            pytest.mark.skip(reason="set methods not implemented in c"),
```

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
It would be nice to add tests to ensure that a neat error is raised for inhomogeneous and non-hashable types.

File: pyccel/codegen/printing/pycode.py (hunk 5 above)
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py:447]: You're correct, I'll include a condition for that

File: pyccel/errors/messages.py (hunk 6 above)
Comment (EmilyBourne) [reply pyccel/errors/messages.py:114]: I am not convinced that tuples and sets are sufficiently similar for tuples to be a drop-in replacement. For example I often use sets to remove duplicate items in a list. How can I do this with a tuple?

Comment (mustapha-belbiad) [reply pyccel/errors/messages.py:114]: I get it now I'm going to change it

File: pyccel/codegen/printing/pycode.py [hunk 8]
```diff
@@ -442,6 +442,12 @@ def _print_PythonList(self, expr):
         args = ', '.join(self._print(i) for i in expr.args)
         return '['+args+']'
 
+    def _print_PythonSet(self, expr):
+        args = ', '.join(self._print(i) for i in expr.args)
+        if len(expr.args) == 0:
+            args += ','
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py:448]: /bot accept
Can't be tested before #1658 or #1659 is fixed

[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py:448]: /bot accept
print an empty set: `{,}`

File: pyccel/parser/semantic.py [hunk 9]
```diff
@@ -2198,6 +2198,18 @@ def _visit_PythonList(self, expr):
...
+            if not message :
+                message = PYCCEL_RESTRICTION_INHOMOG_SET
```
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/parser/semantic.py:2208]: /bot accept
The purpose of this is to catch error messages

File: pyccel/codegen/printing/pycode.py (hunk 5 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py]: Has this been done now? Please provide the commit hash

File: pyccel/ast/builtin_methods/set_methods.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    _shape = None
+    _order = None
+    _rank = 0
+    _precision = None
+    _class_type = NativeVoid()
+    name = 'add'
+
+    def __init__(self, set_variable, new_elem) -> None:
+        if (new_elem.rank > 0 and not isinstance(new_elem, PythonTuple)) :
+            raise TypeError("Can't add an unhashable type")
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/builtin_methods/set_methods.py]: As per the review summary message:
> It would be nice to add tests to ensure that a neat error is raised for inhomogeneous and non-hashable types.

If you have changed something so that the error is raised at the set instantiation please add a test for this

File: pyccel/codegen/printing/pycode.py (hunk 5 above)
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py]: fixed : 5c0ec34

File: pyccel/ast/builtin_methods/set_methods.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+__all__ = ('SetAdd',)
+
+class SetAdd(PyccelInternalFunction) :
+    """
+    Represents a call to the .add() method.
+    
+    Represents a call to the .add() method which add an element
+    to the set if it's not already present.
+    If the element is already in the set, the set remains unchanged.
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_methods/set_methods.py:23]:
```suggestion
    Represents a call to the .add() method which adds an element
    to the set if it's not already present.
    If the element is already in the set, the set remains unchanged.
```

Comment (mustapha-belbiad) [reply pyccel/ast/builtin_methods/set_methods.py]: fixed: 4383428

File: pyccel/parser/semantic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:2208]: I agree with the bot. Line 2208 will never be used because you always describe the error in your TypeError message. Is this line useful?

**Review by EmilyBourne — APPROVED** [review]
Looks good to me. Well done

File: pyccel/codegen/printing/pycode.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/codegen/printing/pycode.py:448]: @mustapha-belbiad Have you tried to run this in Python?
``` python
s = {,}
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py:448]: :facepalm: I think I may have suggested this. Sorry
image

[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/codegen/printing/pycode.py:448]: Unfortunately there is no way to create an empty literal set in Python. One needs to call the constructor explicitly:
``` python
s = set()
```
@mustapha-belbiad Please add a unit test for an empty set.

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py:448]: As mentioned above, I don't think it is possible to add a unit test for an empty set until #1658 or #1659 is fixed as we would need a type annotation to determine the type of the variable.

[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/codegen/printing/pycode.py:448]: The two issues you mention are for Fortran and C. Why are they relevant here?

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py:448]: They are the issues that we have which talk about adding support for variable declarations via type annotations

[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py:448]: I will correct it and convert it to `set()`.

[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py]: fixed : d8f8df1

File: pyccel/codegen/printing/pycode.py [hunk 12]
```diff
@@ -442,6 +442,12 @@ def _print_PythonList(self, expr):
         args = ', '.join(self._print(i) for i in expr.args)
         return '['+args+']'
 
+    def _print_PythonSet(self, expr):
+        args = ', '.join(self._print(i) for i in expr.args)
+        if len(expr.args) == 0:
+            return 'set()'
+        return '{'+args+'}'
```
Comment (yguclu) [inline pyccel/codegen/printing/pycode.py:449]: Should you not place the `if` condition before computing the `args` string?

Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py:449]: Sure, could you please tell me why?

Comment (yguclu) [reply pyccel/codegen/printing/pycode.py:449]: Because there is no need to compute `args` if the length of `expr.args` is 0.

File: pyccel/codegen/printing/pycode.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/codegen/printing/pycode.py]: > As mentioned above, I don't think it is possible to add a unit test for an empty set until #1658 or #1659 is fixed as we would need a type annotation to determine the type of the variable.

I have finally understood what you mean, you were referring to the type of the object contained in the set. All right!

File: pyccel/codegen/printing/pycode.py (hunk 12 above)
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py]: fixed : 2ed79da

File: pyccel/codegen/printing/pycode.py [hunk 13]
```diff
@@ -442,6 +442,12 @@ def _print_PythonList(self, expr):
         args = ', '.join(self._print(i) for i in expr.args)
         return '['+args+']'
 
+    def _print_PythonSet(self, expr):
+        args = ', '.join(self._print(i) for i in expr.args)
+        if len(expr.args) == 0:
+            return 'set()'
```
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py:447]: /bot accept

File: tests/epyccel/test_epyccel_sets.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+            pytest.mark.c]),
+        pytest.param("python", marks = pytest.mark.python)
+    ],
+    scope = "module"
+)
+def language(request):
+    return request.param
+
+def test_add_1(language) :
+    def add_int():
```
Comment (yguclu) [inline tests/epyccel/test_epyccel_sets.py:20]: We should give a better name to the test functions. Here I would call it `test_add_literal_int`.

File: tests/epyccel/test_epyccel_sets.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        a = {1,3,45}
+        a.add(4)
+        return a
+    epyc_add_element = epyccel(add_int, language = language)
+    pyccel_result = epyc_add_element()
+    python_result = add_int()
+    assert python_result == pyccel_result
+
+def test_add_element2(language) :
+    def add_complex():
```
Comment (yguclu) [inline tests/epyccel/test_epyccel_sets.py:30]: This I would call `test_add_literal_complex`.

File: tests/epyccel/test_epyccel_sets.py
Comment (yguclu) [inline tests/epyccel/test_epyccel_sets.py:1]: Please add an example where a set is created using the `set()` function.

By the way, all these test are for scalar numeric types. Do we also support something more complicated, like a set of tuples?

Comment (yguclu) [inline tests/epyccel/test_epyccel_sets.py:1]: And how about constructor calls using iterables? These should work:
``` python
s1 = set(range(5)) # {0, 1, 2, 3, 4}
s2 = {*range(5)}   # {0, 1, 2, 3, 4}

l = [0, 1, 2, 2, 2]

s3 = set(l) # {0, 1, 2}
s4 = {*l}   # {0, 1, 2}
```
Please note that converting a list to a set (and possibly back to a list) is a very common use case in Python.

File: tests/errors/semantic/blocking/SET_UNHASHABLE_TYPE_1.py [hunk 16]
```diff
@@ -0,0 +1,3 @@
+# pylint: disable=missing-function-docstring, missing-module-docstring
+
+l = {(1,2,3)}
```
Comment (yguclu) [inline tests/errors/semantic/blocking/SET_UNHASHABLE_TYPE_1.py:3]: A tuple of integers is definitely hashable.

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py [hunk 17]
```diff
@@ -0,0 +1,6 @@
+# pylint: disable=missing-function-docstring, missing-module-docstring
+
+a = {1,2,3}
+b = [4,5]
+a.add(b)
```
Comment (yguclu) [inline tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py:5]: Before worrying about hashability, this is a problem of incompatible types! The set contains integers.

**Review by yguclu — CHANGES_REQUESTED** [review]
Good job! I have some comments though

File: tests/epyccel/test_epyccel_sets.py
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_sets.py:1]: This is out of scope for this PR. The `set` function has different behaviour to the `{}` operator and is not entirely trivial to handle. We saw this when dealing with tuples:
```python
In [1]: set(1,2,3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-35c708cc93ca> in <module>
----> 1 set(1,2,3)

TypeError: set expected at most 1 argument, got 3

In [2]: {1,2,3}
Out [2]: {1, 2, 3}
```

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py (hunk 17 above)
Comment (mustapha-belbiad) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py:5]: The idea is to attempt adding a hashable type to test for errors in the pop method.

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py [hunk 18]
```diff
@@ -0,0 +1,6 @@
+# pylint: disable=missing-function-docstring, missing-module-docstring
+
+a = {(0,7),(4,2), (9,6)}
+b = 4
+a.add(b)
```
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py:5]: In this case, the error will be raised on line 3.

File: tests/epyccel/test_epyccel_sets.py (hunk 14 above)
Comment (mustapha-belbiad) [reply tests/epyccel/test_epyccel_sets.py]: fixed : 0ff444f

File: tests/epyccel/test_epyccel_sets.py (hunk 15 above)
Comment (mustapha-belbiad) [reply tests/epyccel/test_epyccel_sets.py]: fixed : 0ff444f

File: tests/epyccel/test_epyccel_sets.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+        a = {6j,7j,8j}
+        a.add(9j)
+        return a
+    epyc_add_element = epyccel(add_complex, language = language)
+    pyccel_result = epyc_add_element()
+    python_result = add_complex()
+    assert python_result == pyccel_result
+
+def test_add_element_3(language):
+    def add_element_range():
```
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply tests/epyccel/test_epyccel_sets.py]: fixed : 0ff444f

File: tests/epyccel/test_epyccel_sets.py
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_sets.py:1]: @mustapha-belbiad Please update your PR description and CHANGELOG to make it clear that support is added for the `{}` operator for **sets of scalars**. Currently the PR description makes no note of the restrictions imposed by your implementation

Comment (yguclu) [reply tests/epyccel/test_epyccel_sets.py:1]: It seems to me that being able to create a set using `set(iterable)` should be the priority, before worrying about its methods. The same goes for `list(iterable)` and `tuple(iterable)`. Anyway I agree, this is beyond the scope of this PR.

@mustapha-belbiad Please clarify the limitations of the current implementation in the PR description.

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py (hunk 17 above)
Comment (yguclu) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py:5]: Sorry, I cannot understand what you mean. What has the "pop" method to do with this PR?

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py:5]: What is the point of lines 4 and 5 then? They are just confusing and I suggest to remove them.

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py (hunk 17 above)
Comment (mustapha-belbiad) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py:5]: > Sorry, I cannot understand what you mean. What has the "pop" method to do with this PR?

sorry, I was  talking about the add Method.

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py:5]: what about this example :
```suggestion
a = {4, 7, 9}
b = {3,5}
a.add(b)
```
the point is to test the add method when adding an unhashable type.

File: tests/epyccel/test_epyccel_sets.py
Comment (mustapha-belbiad) [reply tests/epyccel/test_epyccel_sets.py:1]: fixed : 61bb8ff

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py:5]: In your original code you were not testing the add method, because an error was raised on line 3 when you constructed the set.

The point is **not** testing the add method when adding an unhashable type, because that makes really no sense. Let me explain:

1. Pyccel can only create sets from hashable types (or at least some specific hashable types). If you pass an unhashable type to `{}` an error is raised.
2. When using `add` you must pass a single object of the same type as the objects contained in the set. If you pass the wrong type a `TypeError` should be raised. It does not matter if the passed type is hashable or not. There is one correct type and all the others are wrong.

Do you understand what I mean?

File: CHANGELOG.md [hunk 20]
```diff
@@ -11,7 +11,7 @@ All notable changes to this project will be documented in this file.
 -   #1691 : Add Python support for list method `clear()`
 -   #1575 : Add support for homogeneous tuple type annotations on variables.
 -   #1425 : Add support for `numpy.isnan`, `numpy.isinf` and `numpy.isfinite`.
--   #1738 : Add Python support for set method `add()`
+-   #1738 : Add Python support for scalar sets with `{}` and `add()` method.
```
Comment (yguclu) [inline CHANGELOG.md:14]: Can we split into two points? One about constructing sets, the other about the `add` method.

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py (hunk 17 above)
Comment (yguclu) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_1.py:5]: In any case, the test is for the constructor, not for the `add` method. I explained in another thread.

File: CHANGELOG.md (hunk 20 above)
Comment (yguclu) [reply CHANGELOG.md:14]: Yes, but you have a typo: list -> set

Comment (mustapha-belbiad) [reply CHANGELOG.md:14]: Sorry, I will resolve it

Comment (mustapha-belbiad) [reply CHANGELOG.md]: fixed : 61bb8ff

File: tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mustapha-belbiad) [reply tests/errors/semantic/blocking/ADD_SET_UNHASHABLE_TYPE_2.py]: fixed : e441c23

**Review by yguclu — APPROVED** [review]
